File: android/lib/core/services/capture_policy.dart
import 'dart:math' as math;
import 'dart:typed_data';

import '../decode/diagnostics.dart';
import 'decode_isolate.dart';

enum ScanHint { none, moveCloser, moveBack, holdStill, adjustAngle }

enum LockAction { none, lock, unlock }

/// Camera acquisition policy (spec §8): lock focus/exposure once a barcode is
/// located, unlock after [unlockAfterMs] without one; derive a user hint from
/// the finder module size, corner motion and the decode status.
class CapturePolicy {
  final double minModulePx;
  final double maxModulePx;
  final double motionPx;
  final int unlockAfterMs;

  bool _locked = false;
  int? _lastLocatedMs;
  Float64List? _lastCorners;

  CapturePolicy({this.minModulePx = 6, this.maxModulePx = 40, this.motionPx = 10, this.unlockAfterMs = 2000});

  bool get locked => _locked;

  void reset() {
    _locked = false;
    _lastLocatedMs = null;
    _lastCorners = null;
  }

  (ScanHint, LockAction) update(FrameOutcome o, int nowMs) {
    final located = o.corners != null && (o.status == DecodeStatus.ok || o.status == DecodeStatus.rsFailed || o.status == DecodeStatus.badHeader || o.status == DecodeStatus.unsupportedGrid);
    var action = LockAction.none;
    var hint = ScanHint.none;
    if (located) {
      _lastLocatedMs = nowMs;
      if (!_locked) {
        _locked = true;
        action = LockAction.lock;
      }
      if (o.module < minModulePx) {
        hint = ScanHint.moveCloser;
      } else if (o.module > maxModulePx) {
        hint = ScanHint.moveBack;
      } else if (_lastCorners != null && _motion(_lastCorners!, o.corners!) > motionPx) {
        hint = ScanHint.holdStill;
      } else if (o.status == DecodeStatus.rsFailed) {
        hint = ScanHint.adjustAngle;
      }
      _lastCorners = o.corners;
    } else {
      _lastCorners = null;
      if (_locked && _lastLocatedMs != null && nowMs - _lastLocatedMs! >= unlockAfterMs) {
        _locked = false;
        action = LockAction.unlock;
      }
    }
    return (hint, action);
  }

  static double _motion(Float64List a, Float64List b) {
    var worst = 0.0;
    for (var i = 0; i < 8; i += 2) {
      final dx = a[i] - b[i], dy = a[i + 1] - b[i + 1];
      final d = dx * dx + dy * dy;
      if (d > worst) worst = d;
    }
    return worst == 0 ? 0 : math.sqrt(worst);
  }
}

File: web-app/capture-policy.js
/**
 * capture-policy.js — camera acquisition policy for live scan (spec §8).
 *
 * Transliteration of app/lib/core/services/capture_policy.dart: lock focus
 * and exposure once a barcode is located, unlock after unlockAfterMs without
 * one, and derive a user hint from the finder module size, corner motion and
 * the decode status. Constants are Android's, verbatim.
 *
 * One web-only addition: 'tooSmall' counts as located. It is the photo
 * decoder's module floor, which runs after the locator filled corners and
 * module, so a small barcode yields moveCloser through the module rule.
 *
 * Pure logic, no DOM. IIFE; exposes window.CimbarCapturePolicy / module.exports.
 */
'use strict';
(function () {
const isNode = typeof module !== 'undefined' && module.exports;

const LOCATED_STATUSES = ['ok', 'rsFailed', 'badHeader', 'unsupportedGrid', 'tooSmall'];

function isLocated(o) {
  return !!(o && o.corners) && LOCATED_STATUSES.indexOf(o.status) >= 0;
}

function motion(a, b) {
  let worst = 0;
  for (let i = 0; i < 8; i += 2) {
    const dx = a[i] - b[i], dy = a[i + 1] - b[i + 1];
    const d = dx * dx + dy * dy;
    if (d > worst) worst = d;
  }
  return worst === 0 ? 0 : Math.sqrt(worst);
}

class CapturePolicy {
  constructor(opts) {
    const p = opts || {};
    this.minModulePx = p.minModulePx !== undefined ? p.minModulePx : 6;
    this.maxModulePx = p.maxModulePx !== undefined ? p.maxModulePx : 40;
    this.motionPx = p.motionPx !== undefined ? p.motionPx : 10;
    this.unlockAfterMs = p.unlockAfterMs !== undefined ? p.unlockAfterMs : 2000;
    this.reset();
  }

  get locked() { return this._locked; }

  reset() {
    this._locked = false;
    this._lastLocatedMs = null;
    this._lastCorners = null;
  }

  update(o, nowMs) {
    let lockAction = 'none';
    let hint = 'none';
    if (isLocated(o)) {
      this._lastLocatedMs = nowMs;
      if (!this._locked) {
        this._locked = true;
        lockAction = 'lock';
      }
      if (o.module < this.minModulePx) {
        hint = 'moveCloser';
      } else if (o.module > this.maxModulePx) {
        hint = 'moveBack';
      } else if (this._lastCorners !== null && motion(this._lastCorners, o.corners) > this.motionPx) {
        hint = 'holdStill';
      } else if (o.status === 'rsFailed') {
        hint = 'adjustAngle';
      }
      this._lastCorners = o.corners;
    } else {
      this._lastCorners = null;
      if (this._locked && this._lastLocatedMs !== null && nowMs - this._lastLocatedMs >= this.unlockAfterMs) {
        this._locked = false;
        lockAction = 'unlock';
      }
    }
    return { hint, lockAction };
  }
}

const API = { CapturePolicy, isLocated, LOCATED_STATUSES };
if (isNode) module.exports = API; else window.CimbarCapturePolicy = API;
})();
